From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Subject: s390/3270: fix view reference counting
Patch-mainline: v4.7-rc1
Git-commit: b512353c15fe7665893e92e0c989c088505a37b8
References: bnc#980892, LTC#141736

Description:  kernel: fix disconnect/reconnect of 3270 terminals
Symptom:      The disconnect and reconnect of a 3270 terminal is not
              working as expected, the terminal session is not hung
              up and the new size of the reconnected terminal is not
              detected.
Problem:      The 3270 driver does not have support for disconnected
              terminals.
Solution:     Add support for disconnected 3270 terminals.
Reproduction: Open a 3270 terminal, login on the tty, disconnected the
              terminal and reconnect with a terminal of different size.

Upstream-Description:

              s390/3270: fix view reference counting

              The tty3270_close function set tty->driver_data to NULL after the
              open count dropped to zero but tty3270_cleanup still needs the
              pointer to the tty3270 view to drop the reference count.

              Move the clearing of tty->driver_data to tty3270_cleanup and
              drop the unnecessary tty_port_tty_set calls in tty3270_install.

              Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>


Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: John Jolly <jjolly@suse.de>
---
 drivers/s390/char/tty3270.c |   14 ++++++--------
 1 file changed, 6 insertions(+), 8 deletions(-)

--- a/drivers/s390/char/tty3270.c
+++ b/drivers/s390/char/tty3270.c
@@ -924,10 +924,8 @@ static int tty3270_install(struct tty_dr
 		tty->winsize.ws_row = tp->view.rows - 2;
 		tty->winsize.ws_col = tp->view.cols;
 		tp->port.low_latency = 0;
-		/* why to reassign? */
-		tty_port_tty_set(&tp->port, tty);
 		tp->inattr = TF_INPUT;
-		return tty_port_install(&tp->port, driver, tty);
+		goto port_install;
 	}
 	if (tty3270_max_index < tty->index + 1)
 		tty3270_max_index = tty->index + 1;
@@ -953,7 +951,6 @@ static int tty3270_install(struct tty_dr
 		return rc;
 	}
 
-	tty_port_tty_set(&tp->port, tty);
 	tp->port.low_latency = 0;
 	tty->winsize.ws_row = tp->view.rows - 2;
 	tty->winsize.ws_col = tp->view.cols;
@@ -975,6 +972,7 @@ static int tty3270_install(struct tty_dr
 
 	raw3270_activate_view(&tp->view);
 
+port_install:
 	rc = tty_port_install(&tp->port, driver, tty);
 	if (rc) {
 		raw3270_put_view(&tp->view);
@@ -1011,18 +1009,18 @@ tty3270_close(struct tty_struct *tty, st
 
 	if (tty->count > 1)
 		return;
-	if (tp) {
-		tty->driver_data = NULL;
+	if (tp)
 		tty_port_tty_set(&tp->port, NULL);
-	}
 }
 
 static void tty3270_cleanup(struct tty_struct *tty)
 {
 	struct tty3270 *tp = tty->driver_data;
 
-	if (tp)
+	if (tp) {
+		tty->driver_data = NULL;
 		raw3270_put_view(&tp->view);
+	}
 }
 
 /*
